### DISCUSSION TITLE
How pnpm judge if  module is exist in store

### DISCUSSION THREAD

**Discussion by guojingkang** [pnpm/pnpm#4126] [Q&A] [answered]
hi, i am learning pnpm。i find the module files are stored in /Users/.pnpm-store/v3/files, and use the function to get file path 
```js

function contentPathFromIntegrity(
    integrity,
    fileType,
) {
    const sri = ssri.parse(integrity, { single: true });
    return contentPathFromHex(fileType, sri.hexDigest());
}

function contentPathFromHex(fileType, hex) {
    const p = path.join(hex.slice(0, 2), hex.slice(2));
    switch (fileType) {
        case 'exec':
            return `${p}-exec`;
        case 'nonexec':
            return p;
        case 'index':
            return `${p}-index.json`;
    }
}
```
i know that  a module xxx-index.json is generated from module's integrity with the function above, and with the index.json, i can find other file。 but i dont know how pnpm judge if the module exist in the store when i install or add 。

**Comment by zkochan** [comment, marked as answer]
The name of the index file (`-index.json`) contains the package's checksum. So we find the index file using the package's integrity checksum.

**Reply by guojingkang** [reply]
thank's ！
